src/Lecture4.hs

```diff
@@ -135,7 +141,37 @@ errors. We will simply return an optional result here.
 -}
 
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow r = case wordsBy ',' r of
+   [p, t, c] -> do
+      p' <- parseProduct p
+      t' <- parseTradeType $ trim t
+      c' <- parseCost $ trim c
+      pure $ Row p' t' c'
+   _ -> Nothing
+
+wordsBy :: Char -> String -> [String]
+wordsBy c s = case dropWhile (== c) s of
+   "" -> []
+   s' -> w : wordsBy c s''
+      where (w, s'') = break (== c) s'
+
+trim :: String -> String
+trim = dropWhileEnd isSpace . dropWhile isSpace
+
+parseProduct :: String -> Maybe String
+parseProduct "" = Nothing
+parseProduct s = Just s
+
+parseTradeType :: String -> Maybe TradeType
+parseTradeType "Buy" = Just Buy
+parseTradeType "Sell" = Just Sell
+parseTradeType _ = Nothing
```

Comment:
`TradeType` already derives the `Read` instance so this function is just `readMaybe` 👌🏻